README
Pg-PQ version
=============

This Perl package wraps PostgreSQL libpq library.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

  libpq
  Test::postgresql for testing

COPYRIGHT AND LICENCE

Copyright (C) 2011-2013 by Qindel Formacion y Servicios S. L.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10.1 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Pg::PQ.

0.15  Oct 10, 2017
        - Fix version libpq version detection for PostgreSQL 10 (bug
          report and patch by Lee Aylward).
        - Avoid empty options in connection string (bug report and
          patch by Ivan Panchenko).
        - Add more tests.
0.14  Oct 28, 2014
        - use Test::PostgreSQL for testing

0.13  Jun 21, 2013
        - rowAsHash and rowsAsHashes were handling NULL values
          incorrectly, resulting in "Modification of non-creatable
          hash value attempted" errors
        - wrap PQconndefaults
        - fix POD errors

0.12  Apr 16, 2013
        - fix error in rowsAsHashes that could result in segmentation
          fault in some cases

0.11  Apr 9, 2013
        - add support for rowAsHash and rowsAsHashes methods

0.10  Feb 1, 2013
        - Makefile.PL LIB parameters were in the wrong order (bug
          report and solution by Geoffrey D. Bennett).
        - nonBlocking() was broken

0.09  Jul 2, 2012
        - return libpq version as obtained from pg_config instead of
          from libVersion

0.08  May 10, 2012
        - wrap libVersion
        - set client_encoding to UTF8 on new connections
        - allow compiling against development versions of PostgreSQL
          (bug report by erix@perlmonks)

0.07  May 9, 2012
        - assume database encoding is UTF8
        - better handling of errorDescription and errorField
        - use Test::postgresql for testing
        - add support for payload field in notifies (bug report by
          Mischa Spiegelmock)

0.06  Jan 3, 2012
        - allow compilation with older versions of libpq
        - OUTPUT directive was missing from PQcmdTuples wrapper

0.05  Jul 18, 2011
        - accept a hash ref as conninfo
        - start was calling PQconnectdb instead of PQconnectStart
          under the hood

0.04  Jun 13, 2011
	- add conditional code to leave out methods not available from
          libpq 8.x.x when this version of the client library is used

0.03  Jun 9, 2011
        - several doc improvements
        - include typemap file in distribution

0.02
        - several doc improvements

0.01  Mon May 30 16:54:48 2011
	- original version; created by h2xs 1.23 with options
		-An Pg::PQ



enums.c
/*
 *
 * This file is generated by gen_constants.pl
 * Do not edit by hand!
 *
 */

SV *enum2sv_ConnStatusType[9];
SV *enum2sv_CopyResultFlag[9];
SV *enum2sv_ExecStatusType[8];
SV *enum2sv_PGTransactionStatusType[5];
SV *enum2sv_PGVerbosity[3];
SV *enum2sv_PostgresPollingStatusType[5];

static void
init_constants(void) {
    enum2sv_ConnStatusType[0] = make_constant("CONNECTION_OK", 13, CONNECTION_OK);
    enum2sv_ConnStatusType[1] = make_constant("CONNECTION_BAD", 14, CONNECTION_BAD);
    enum2sv_ConnStatusType[2] = make_constant("CONNECTION_STARTED", 18, CONNECTION_STARTED);
    enum2sv_ConnStatusType[3] = make_constant("CONNECTION_MADE", 15, CONNECTION_MADE);
    enum2sv_ConnStatusType[4] = make_constant("CONNECTION_AWAITING_RESPONSE", 28, CONNECTION_AWAITING_RESPONSE);
    enum2sv_ConnStatusType[5] = make_constant("CONNECTION_AUTH_OK", 18, CONNECTION_AUTH_OK);
    enum2sv_ConnStatusType[6] = make_constant("CONNECTION_SETENV", 17, CONNECTION_SETENV);
    enum2sv_ConnStatusType[7] = make_constant("CONNECTION_SSL_STARTUP", 22, CONNECTION_SSL_STARTUP);
    enum2sv_ConnStatusType[8] = make_constant("CONNECTION_NEEDED", 17, CONNECTION_NEEDED);

    enum2sv_CopyResultFlag[0] = make_constant("COPYRESULTFLAG_0", 16, 0);
    enum2sv_CopyResultFlag[1] = make_constant("PG_COPYRES_ATTRS", 16, PG_COPYRES_ATTRS);
    enum2sv_CopyResultFlag[2] = make_constant("PG_COPYRES_TUPLES", 17, PG_COPYRES_TUPLES);
    enum2sv_CopyResultFlag[3] = make_constant("COPYRESULTFLAG_3", 16, 3);
    enum2sv_CopyResultFlag[4] = make_constant("PG_COPYRES_EVENTS", 17, PG_COPYRES_EVENTS);
    enum2sv_CopyResultFlag[5] = make_constant("COPYRESULTFLAG_5", 16, 5);
    enum2sv_CopyResultFlag[6] = make_constant("COPYRESULTFLAG_6", 16, 6);
    enum2sv_CopyResultFlag[7] = make_constant("COPYRESULTFLAG_7", 16, 7);
    enum2sv_CopyResultFlag[8] = make_constant("PG_COPYRES_NOTICEHOOKS", 22, PG_COPYRES_NOTICEHOOKS);

    enum2sv_ExecStatusType[0] = make_constant("PGRES_EMPTY_QUERY", 17, PGRES_EMPTY_QUERY);
    enum2sv_ExecStatusType[1] = make_constant("PGRES_COMMAND_OK", 16, PGRES_COMMAND_OK);
    enum2sv_ExecStatusType[2] = make_constant("PGRES_TUPLES_OK", 15, PGRES_TUPLES_OK);
    enum2sv_ExecStatusType[3] = make_constant("PGRES_COPY_OUT", 14, PGRES_COPY_OUT);
    enum2sv_ExecStatusType[4] = make_constant("PGRES_COPY_IN", 13, PGRES_COPY_IN);
    enum2sv_ExecStatusType[5] = make_constant("PGRES_BAD_RESPONSE", 18, PGRES_BAD_RESPONSE);
    enum2sv_ExecStatusType[6] = make_constant("PGRES_NONFATAL_ERROR", 20, PGRES_NONFATAL_ERROR);
    enum2sv_ExecStatusType[7] = make_constant("PGRES_FATAL_ERROR", 17, PGRES_FATAL_ERROR);

    enum2sv_PGTransactionStatusType[0] = make_constant("PQTRANS_IDLE", 12, PQTRANS_IDLE);
    enum2sv_PGTransactionStatusType[1] = make_constant("PQTRANS_ACTIVE", 14, PQTRANS_ACTIVE);
    enum2sv_PGTransactionStatusType[2] = make_constant("PQTRANS_INTRANS", 15, PQTRANS_INTRANS);
    enum2sv_PGTransactionStatusType[3] = make_constant("PQTRANS_INERROR", 15, PQTRANS_INERROR);
    enum2sv_PGTransactionStatusType[4] = make_constant("PQTRANS_UNKNOWN", 15, PQTRANS_UNKNOWN);

    enum2sv_PGVerbosity[0] = make_constant("PQERRORS_TERSE", 14, PQERRORS_TERSE);
    enum2sv_PGVerbosity[1] = make_constant("PQERRORS_DEFAULT", 16, PQERRORS_DEFAULT);
    enum2sv_PGVerbosity[2] = make_constant("PQERRORS_VERBOSE", 16, PQERRORS_VERBOSE);

    enum2sv_PostgresPollingStatusType[0] = make_constant("PGRES_POLLING_FAILED", 20, PGRES_POLLING_FAILED);
    enum2sv_PostgresPollingStatusType[1] = make_constant("PGRES_POLLING_READING", 21, PGRES_POLLING_READING);
    enum2sv_PostgresPollingStatusType[2] = make_constant("PGRES_POLLING_WRITING", 21, PGRES_POLLING_WRITING);
    enum2sv_PostgresPollingStatusType[3] = make_constant("PGRES_POLLING_OK", 16, PGRES_POLLING_OK);
    enum2sv_PostgresPollingStatusType[4] = make_constant("PGRES_POLLING_ACTIVE", 20, PGRES_POLLING_ACTIVE);

}
static SV *
ConnStatusType2sv(I32 ix) {
    SV *sv;
    ix -= CONNECTION_OK;
    if ((ix < 0) || (ix >= 9)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_ConnStatusType[ix]);
    return sv;
}

static SV *
CopyResultFlag2sv(I32 ix) {
    SV *sv;
    ix -= PG_COPYRES_ATTRS - 1;
    if ((ix < 0) || (ix >= 9)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_CopyResultFlag[ix]);
    return sv;
}

static SV *
ExecStatusType2sv(I32 ix) {
    SV *sv;
    ix -= PGRES_EMPTY_QUERY;
    if ((ix < 0) || (ix >= 8)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_ExecStatusType[ix]);
    return sv;
}

static SV *
PGTransactionStatusType2sv(I32 ix) {
    SV *sv;
    ix -= PQTRANS_IDLE;
    if ((ix < 0) || (ix >= 5)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_PGTransactionStatusType[ix]);
    return sv;
}

static SV *
PGVerbosity2sv(I32 ix) {
    SV *sv;
    ix -= PQERRORS_TERSE;
    if ((ix < 0) || (ix >= 3)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_PGVerbosity[ix]);
    return sv;
}

static SV *
PostgresPollingStatusType2sv(I32 ix) {
    SV *sv;
    ix -= PGRES_POLLING_FAILED;
    if ((ix < 0) || (ix >= 5)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_PostgresPollingStatusType[ix]);
    return sv;
}



Makefile.PL

use ExtUtils::MakeMaker;

use strict;
use warnings;

use 5.010;

my ($version, $major) = `pg_config --version 2>&1` =~ /PostgreSQL\s+((\d+)\.\d+(?:\.\d+|\w+))\b/i;
unless (defined $major and $major >= 8) {
    die "libpq 8.x.x or later required (" . ($version // 'none') . " found)\n";
}

print STDERR "PostgreSQL $version found\n";

my $incdir = `pg_config --includedir`;
my $libdir = `pg_config --libdir`;
chomp($incdir, $libdir);

my %prereq_pm;
$prereq_pm{'Test::postgresql'} = '0.09' if $ENV{AUTOMATED_TESTING};

WriteMakefile( NAME          => 'Pg::PQ',
               VERSION_FROM  => 'lib/Pg/PQ.pm',
               PREREQ_PM     => \%prereq_pm,
               ABSTRACT_FROM => 'lib/Pg/PQ.pm',
               AUTHOR        => 'Salvador Fandino <sfandino@yahoo.com>',
               LIBS          => ["-L$libdir -lpq"],
               DEFINE        => qq(-DPQMAJOR=$major -DPQVERSION='"$version"'), # e.g., '-DHAVE_SOMETHING'
               INC           => "-I. -I$incdir",
               # OBJECT        => '$(O_FILES)', # link all the C files too
             );


build_tools/gen_constants.pl
#!/usr/bin/perl

use strict;
use warnings;

open E, "< enums.decl" or die "unable to open enums.h";
open O, "> enums.h" or die "unable to open enums.c";

my %enum;
my %tag;
my %ver;
my @s;
my $c;
my $last = 0;
while(<E>) {
    s|/\*.*/||;
    s|^\s\* .*||;
    s|^.*\*/||;
    if (my ($name, $tag, $ver) = /^enum\s*(\w+)\s*(?:=>\s*(\w+))?(?:\s+if\s+(\d+\.(?:\d+(?:\.\d+)?)?))?/) {
        $last = 0;
	# $name =~ s/^ldap_//;
	$c = $enum{$name} = [];
        $tag{$name} = $tag;
        if (defined $ver) {
            $ver =~ /^(\d+)(?:\.(\d+)(?:\.(\d+))?)?$/ or die "bad version number $ver";
            $ver{$name} = sprintf("#if PG_VERSION_NUM >= %d%02d%02d\n", $1, $2 || 0, $3 || 0)
        }
    }
    elsif (/^\s*((?:P[QG]|CONN)\w+)\s*=\s*(\d+),/) {
        $c->[$2] = $1;

        $last = $2 + 1;
    }
    elsif (/^\s*((?:P[QG]|CONN)\w+)\s*,/) {
        $c->[$last++] = $1;
    }
    elsif (m|^\s*s/(\w+)/(\w*)/|) {
        push @s, [$1, $2];
    }
}

@s = sort { length($b->[0]) <=> length($a->[0]) } @s;
# use Data::Dumper; print Dumper \@s;

print O <<HEAD;
/*
 *
 * This file is generated by gen_constants.pl
 * Do not edit by hand!
 *
 */

HEAD

for my $enum (sort keys %enum) {
    my $ver = $ver{$enum};
    defined $ver and print O $ver;
    my $len = @{$enum{$enum}};
    print O "SV *enum2sv_${enum}[$len];\n";
    defined $ver and print O "#endif\n";
}

# print O <<DECL for @c;
# SV *${_}_sv;
# DECL

print O <<INIT;

static void
init_constants(void) {
INIT

for my $enum (sort keys %enum) {
    my $c = $enum{$enum};
    my $tag = $tag{$enum};
    my $ver = $ver{$enum};

    defined $ver and print O $ver;

    if (defined $tag and length $tag) {
        $tag = qq("$tag");
    }
    else {
        $tag = 'NULL';
    }


    for my $ix (0..$#$c) {
	my $name = $c->[$ix];
	my $value;
	if (defined $name) {
	    $value = $name;
            unless (defined $c->[0]) {
                $c->[0] = "$value - $ix"
            }
	}
	else {
	    $name = uc "${enum}_$ix";
	    $value = $ix;
	}

        for my $s (@s) {
            my ($f, $t) = @$s;
            $name =~ s/^$f/$t/;
        }

	printf O <<C, $ix, $name, length($name), $value, $tag;
    enum2sv_${enum}[%d] = make_constant("%s", %d, %s, %s);
C

    }
    defined $ver and print O "#endif\n";

    print O "\n";
}

print O <<END;
}
END

for my $enum (sort keys %enum) {
    my $ver = $ver{$enum};
    defined $ver and print O $ver;

    my $len = @{$enum{$enum}};
    print O <<ETS;
static SV *
${enum}2sv(I32 ix) {
    SV *sv;
    ix -= $enum{$enum}[0];
    if ((ix < 0) || (ix >= $len)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_${enum}[ix]);
    return sv;
}
ETS
    defined $ver and print O "#endif\n";
    print O "\n";
}


enums.h
/*
 *
 * This file is generated by gen_constants.pl
 * Do not edit by hand!
 *
 */

SV *enum2sv_ConnStatusType[9];
#if PG_VERSION_NUM >= 80400
SV *enum2sv_CopyResultFlag[9];
#endif
SV *enum2sv_ExecStatusType[8];
SV *enum2sv_PGDiag[0];
SV *enum2sv_PGTransactionStatusType[5];
SV *enum2sv_PGVerbosity[3];
SV *enum2sv_PostgresPollingStatusType[5];

static void
init_constants(void) {
    enum2sv_ConnStatusType[0] = make_constant("CONNECTION_OK", 13, CONNECTION_OK, "connection");
    enum2sv_ConnStatusType[1] = make_constant("CONNECTION_BAD", 14, CONNECTION_BAD, "connection");
    enum2sv_ConnStatusType[2] = make_constant("CONNECTION_STARTED", 18, CONNECTION_STARTED, "connection");
    enum2sv_ConnStatusType[3] = make_constant("CONNECTION_MADE", 15, CONNECTION_MADE, "connection");
    enum2sv_ConnStatusType[4] = make_constant("CONNECTION_AWAITING_RESPONSE", 28, CONNECTION_AWAITING_RESPONSE, "connection");
    enum2sv_ConnStatusType[5] = make_constant("CONNECTION_AUTH_OK", 18, CONNECTION_AUTH_OK, "connection");
    enum2sv_ConnStatusType[6] = make_constant("CONNECTION_SETENV", 17, CONNECTION_SETENV, "connection");
    enum2sv_ConnStatusType[7] = make_constant("CONNECTION_SSL_STARTUP", 22, CONNECTION_SSL_STARTUP, "connection");
    enum2sv_ConnStatusType[8] = make_constant("CONNECTION_NEEDED", 17, CONNECTION_NEEDED, "connection");

#if PG_VERSION_NUM >= 80400
    enum2sv_CopyResultFlag[0] = make_constant("PG_COPYRES_0", 12, 0, "copyres");
    enum2sv_CopyResultFlag[1] = make_constant("PG_COPYRES_ATTRS", 16, PG_COPYRES_ATTRS, "copyres");
    enum2sv_CopyResultFlag[2] = make_constant("PG_COPYRES_TUPLES", 17, PG_COPYRES_TUPLES, "copyres");
    enum2sv_CopyResultFlag[3] = make_constant("PG_COPYRES_3", 12, 3, "copyres");
    enum2sv_CopyResultFlag[4] = make_constant("PG_COPYRES_EVENTS", 17, PG_COPYRES_EVENTS, "copyres");
    enum2sv_CopyResultFlag[5] = make_constant("PG_COPYRES_5", 12, 5, "copyres");
    enum2sv_CopyResultFlag[6] = make_constant("PG_COPYRES_6", 12, 6, "copyres");
    enum2sv_CopyResultFlag[7] = make_constant("PG_COPYRES_7", 12, 7, "copyres");
    enum2sv_CopyResultFlag[8] = make_constant("PG_COPYRES_NOTICEHOOKS", 22, PG_COPYRES_NOTICEHOOKS, "copyres");
#endif

    enum2sv_ExecStatusType[0] = make_constant("PGRES_EMPTY_QUERY", 17, PGRES_EMPTY_QUERY, "pgres");
    enum2sv_ExecStatusType[1] = make_constant("PGRES_COMMAND_OK", 16, PGRES_COMMAND_OK, "pgres");
    enum2sv_ExecStatusType[2] = make_constant("PGRES_TUPLES_OK", 15, PGRES_TUPLES_OK, "pgres");
    enum2sv_ExecStatusType[3] = make_constant("PGRES_COPY_OUT", 14, PGRES_COPY_OUT, "pgres");
    enum2sv_ExecStatusType[4] = make_constant("PGRES_COPY_IN", 13, PGRES_COPY_IN, "pgres");
    enum2sv_ExecStatusType[5] = make_constant("PGRES_BAD_RESPONSE", 18, PGRES_BAD_RESPONSE, "pgres");
    enum2sv_ExecStatusType[6] = make_constant("PGRES_NONFATAL_ERROR", 20, PGRES_NONFATAL_ERROR, "pgres");
    enum2sv_ExecStatusType[7] = make_constant("PGRES_FATAL_ERROR", 17, PGRES_FATAL_ERROR, "pgres");


    enum2sv_PGTransactionStatusType[0] = make_constant("PQTRANS_IDLE", 12, PQTRANS_IDLE, "pqtrans");
    enum2sv_PGTransactionStatusType[1] = make_constant("PQTRANS_ACTIVE", 14, PQTRANS_ACTIVE, "pqtrans");
    enum2sv_PGTransactionStatusType[2] = make_constant("PQTRANS_INTRANS", 15, PQTRANS_INTRANS, "pqtrans");
    enum2sv_PGTransactionStatusType[3] = make_constant("PQTRANS_INERROR", 15, PQTRANS_INERROR, "pqtrans");
    enum2sv_PGTransactionStatusType[4] = make_constant("PQTRANS_UNKNOWN", 15, PQTRANS_UNKNOWN, "pqtrans");

    enum2sv_PGVerbosity[0] = make_constant("PQERRORS_TERSE", 14, PQERRORS_TERSE, "pqerrors");
    enum2sv_PGVerbosity[1] = make_constant("PQERRORS_DEFAULT", 16, PQERRORS_DEFAULT, "pqerrors");
    enum2sv_PGVerbosity[2] = make_constant("PQERRORS_VERBOSE", 16, PQERRORS_VERBOSE, "pqerrors");

    enum2sv_PostgresPollingStatusType[0] = make_constant("PGRES_POLLING_FAILED", 20, PGRES_POLLING_FAILED, "pgres_polling");
    enum2sv_PostgresPollingStatusType[1] = make_constant("PGRES_POLLING_READING", 21, PGRES_POLLING_READING, "pgres_polling");
    enum2sv_PostgresPollingStatusType[2] = make_constant("PGRES_POLLING_WRITING", 21, PGRES_POLLING_WRITING, "pgres_polling");
    enum2sv_PostgresPollingStatusType[3] = make_constant("PGRES_POLLING_OK", 16, PGRES_POLLING_OK, "pgres_polling");
    enum2sv_PostgresPollingStatusType[4] = make_constant("PGRES_POLLING_ACTIVE", 20, PGRES_POLLING_ACTIVE, "pgres_polling");

}
static SV *
ConnStatusType2sv(I32 ix) {
    SV *sv;
    ix -= CONNECTION_OK;
    if ((ix < 0) || (ix >= 9)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_ConnStatusType[ix]);
    return sv;
}

#if PG_VERSION_NUM >= 80400
static SV *
CopyResultFlag2sv(I32 ix) {
    SV *sv;
    ix -= PG_COPYRES_ATTRS - 1;
    if ((ix < 0) || (ix >= 9)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_CopyResultFlag[ix]);
    return sv;
}
#endif

static SV *
ExecStatusType2sv(I32 ix) {
    SV *sv;
    ix -= PGRES_EMPTY_QUERY;
    if ((ix < 0) || (ix >= 8)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_ExecStatusType[ix]);
    return sv;
}

static SV *
PGDiag2sv(I32 ix) {
    SV *sv;
    ix -= ;
    if ((ix < 0) || (ix >= 0)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_PGDiag[ix]);
    return sv;
}

static SV *
PGTransactionStatusType2sv(I32 ix) {
    SV *sv;
    ix -= PQTRANS_IDLE;
    if ((ix < 0) || (ix >= 5)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_PGTransactionStatusType[ix]);
    return sv;
}

static SV *
PGVerbosity2sv(I32 ix) {
    SV *sv;
    ix -= PQERRORS_TERSE;
    if ((ix < 0) || (ix >= 3)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_PGVerbosity[ix]);
    return sv;
}

static SV *
PostgresPollingStatusType2sv(I32 ix) {
    SV *sv;
    ix -= PGRES_POLLING_FAILED;
    if ((ix < 0) || (ix >= 5)) {
        return newSViv(ix);
    }
    sv = newSVsv(enum2sv_PostgresPollingStatusType[ix]);
    return sv;
}

